File: AtCoder/ABC/159/cpp/B_String_Palindrome.cpp
#include <bits/stdc++.h>
using namespace std;

string S;

bool is_palindrome(string s, int left, int right) {
  while (left <= right) {
    if (s[left] != s[right]) {
      return false;
    }
    ++left;
    --right;
  }
  return true;
}

int main() {
  cin >> S;
  int N = S.size();
  bool res = is_palindrome(S, 0, N - 1) && is_palindrome(S, 0, (N - 1) / 2 - 1) &&
             is_palindrome(S, (N + 3) / 2 - 1, N - 1);
  if (res) {
    puts("Yes");
  } else {
    puts("No");
  }
}

File: AtCoder/ABC/159/python/B_String_Palindrome.py
S = input()


def is_palindrome(left, right):
    while left <= right:
        if S[left] != S[right]:
            return False
        left += 1
        right -= 1
    return True


N = len(S)
res = is_palindrome(0, N - 1) and is_palindrome(0, (N - 1) // 2 -
                                                1) and is_palindrome((N + 3) // 2 - 1, N - 1)
if res:
    print("Yes")
else:
    print("No")
